persistence/sql/persister_grant_jwk.go

```diff
@@ -0,0 +1,196 @@
+package sql
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gobuffalo/pop/v5"
+	"gopkg.in/square/go-jose.v2"
+
+	"github.com/ory/hydra/oauth2/trust"
+	"github.com/ory/x/errorsx"
+
+	"github.com/ory/x/sqlcon"
+)
+
+var _ trust.GrantManager = &Persister{}
+
+const scopeSeparator = " "
+
+func (p *Persister) CreateGrant(ctx context.Context, g trust.Grant, publicKey jose.JSONWebKey) error {
+	// add key, if it doesn't exist
+	if _, err := p.GetKey(ctx, g.PublicKey.Set, g.PublicKey.KeyID); err != nil {
+		if errorsx.Cause(err) != sqlcon.ErrNoRows {
+			return err
+		}
+
+		if err = p.AddKey(ctx, g.PublicKey.Set, &publicKey); err != nil {
+			return err
+		}
+	}
+
+	data := p.sqlDataFromJWTGrant(g)
+
+	return sqlcon.HandleError(p.Connection(ctx).Create(&data))
+}
+
+func (p *Persister) GetConcreteGrant(ctx context.Context, id string) (trust.Grant, error) {
+	var data trust.SQLData
+	if err := p.Connection(ctx).Where("id = ?", id).First(&data); err != nil {
+		return trust.Grant{}, sqlcon.HandleError(err)
+	}
+
+	return p.jwtGrantFromSQlData(data), nil
+}
+
+func (p *Persister) DeleteGrant(ctx context.Context, id string) error {
+	grant, err := p.GetConcreteGrant(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	return p.transaction(ctx, func(ctx context.Context, c *pop.Connection) error {
+		if err := p.Connection(ctx).Destroy(&trust.SQLData{ID: grant.ID}); err != nil {
+			return sqlcon.HandleError(err)
+		}
+
+		return p.DeleteKey(ctx, grant.PublicKey.Set, grant.PublicKey.KeyID)
+	})
+}
+
+func (p *Persister) GetGrants(ctx context.Context, limit, offset int, optionalIssuer string) ([]trust.Grant, error) {
+	grantsData := make([]trust.SQLData, 0)
+
+	query := p.Connection(ctx).Paginate(offset/limit+1, limit).Order("id")
+	if optionalIssuer != "" {
+		query = query.Where("issuer = ?", optionalIssuer)
```

Comment:
Same as the previous, the database will use the existing `(issuer, subject, key_id)` index. It may not be as efficient as a single-column index on `issuer`, but keeping a single index for the three columns instead of 3 indexes should save storage space.